modules/cmf/src/main/java/org/jpos/cmf/AmountFee.java

```diff
@@ -0,0 +1,122 @@
+        setConversionRate(conversionRate);
+    }
+
+    public AmountFee(FeeType feeType, CMFAmount fee, CMFAmount settlementFee, String conversionRate) {
+        setFeeType(feeType);
+        setFee(fee);
+        setSettlementFee(settlementFee);
+        if (conversionRate == null) throw new IllegalArgumentException("null conversionRate");
+//        calculateConversionRate();
+        setConversionRate(conversionRate);
+    }
+
+    public AmountFee(String data) {
+        if (data.length() != SERIALIZED_DATA_LENGTH) throw new IllegalArgumentException("Invalid data length");
+        feeType = FeeType.fromCode(data.substring(0, 2));
+        fee = new CMFAmount(data.substring(2, 15));
+        setConversionRate(data.substring(15, 23));
+        settlementFee = new CMFAmount(data.substring(23));
+    }
+
+    public FeeType getFeeType() {
+        return feeType;
+    }
+
+    public void setFeeType(FeeType feeType) {
+        this.feeType = requireNonNull(feeType);
+    }
+
+    public CMFAmount getFee() {
+        return fee;
+    }
+
+    public void setFee(CMFAmount fee) {
+        this.fee = requireNonNull(fee);
+    }
+
+    public CMFAmount getSettlementFee() {
+        return settlementFee;
+    }
+
+    public void setSettlementFee(CMFAmount settlementFee) {
+        this.settlementFee = requireNonNull(settlementFee);
+    }
+
+    public BigDecimal getConversionRate() {
+        return conversionRate;
+    }
+
+    public void setConversionRate(BigDecimal conversionRate) {
+        // BBB: what if the value has more than 7 significant digits (precision)?
+        //      should we round the value or reject it?
+        this.conversionRate = requireNonNull(conversionRate);
+    }
+
+    public void setConversionRate(String conversionRate) {
+        if (conversionRate != null) {
+            // Conversion rate is N8, where first digit moves the decimal point to the left
+            // But: ISO8583-1:2003 "6.2.4 Conversion rates"
+            //    If the value of the first digit is 8, a single zero after the decimal point is assumed.
+            //    If the value of the first digit is 9, two zeroes after the decimal point are assumed.
+            int conversionScale = conversionRate.charAt(0) - '0';
+            if (conversionScale == 8 || conversionScale == 9)
+                conversionScale++;
+            setConversionRate(new BigDecimal(new BigInteger(conversionRate.substring(1)), conversionScale));
+        }
+    }
+
+    public String serialize() {
+        // TODO: 1. Correctly serialize for cases with with high decimal counts
+        //       ----
+        //       2. Maybe we should check that the (unscaled) value has no more than 7 digits?
+        //         Should we round, fail?
+        return feeType.getCode() +
+                fee.serialize() +
+                conversionRate.scale() + ISOUtil.zeropad(conversionRate.unscaledValue().intValue(), 7) +
+                settlementFee.serialize();
```

Comment:
I believe we cannot push this as is, so correct TODO 1 is a must.
For TODO 2: yes we should, but on `setConversionRate`

Reply:
Puff! El temita con el first digit 8 o 9 derivó de una mala interpretación mía hace unos meses.
No tiene misterios, no hay que hacer un conversionScale++, no hay un "skip" en ese momento... el texto de la spec v2003 para esos dos casos es simplemente a modo de clarificación y no para sugerir que 8 y 9 deben tratarse de manera especial.

Así que nada, es un corrimiento normal de la coma, tanto al leer como al serializar.

Lo que sí ya hice en `setConversionRate(BigDecimal)` es redondear la precisión a 7 cifras significativas (usando ese `MathContext` que ya tenía definido).